[WIP] Add vim.* to uv thread

erw7 opened the pull request:
Attempt to expose part of `vim.*` to the uv thread using `luv_set_thread_cb()`.

Ref. #15655.

- [x] Add vim.cjson to uv thread.
- [ ]  Add more test.

erw7 committed:
feat(lua): add vim.loop in thread

erw7 committed:
feat(lua): add vim.mpack in thread

erw7 committed:
fix(lua): replace caller of callback in thread

erw7 committed:
feat(lua): add vim.is_thread()

erw7 committed:
feat(lua): replace print() in thread

erw7 committed:
feat(lua): add vim.diff in thread

bfredl commented on src/nvim/lua/executor.c:
```diff
@@ -1709,6 +1824,18 @@ static int regex_match_line(lua_State *lstate)
 
+static int nlua_is_thread(lua_State *lstate)
+{
+  int argc = lua_gettop(lstate);
+  if (argc) {
+    lua_pop(lstate, argc);
+  }
```
not needed.

erw7 replied:
Done.

bfredl commented on src/nvim/lua/executor.c:
```diff
@@ -252,6 +252,52 @@ static int nlua_luv_cfpcall(lua_State *lstate, int nargs, int nresult,
+{
+  char *error = (char *)argv[0];
+  msg_ext_set_kind("lua_error");
+  emsgf_multiline("Error occuring in luv thread:\n%s", error);
```
more specific: "error occuring in luv callback in thread"

justinmk replied:
also don't need "occuring"

```suggestion
  emsgf_multiline("Error in luv callback thread:\n%s", error);
```

erw7 replied:
Done.

bfredl commented on src/nvim/lua/executor.c:
```diff
@@ -624,9 +676,61 @@ void nlua_init(void)
+  lua_setfield(lstate, -2, "luv");
+  lua_pop(lstate, 3);
+
+  // vim.mpack
```
`vim.NIL` and `vim._empty_dict_mt` is also needed for this. Though I'll suggest breaking out a helper function for shared libs (mpack, diff, also `shared.lua` ?) and move these there instead of duplicating more code.

erw7 replied:
Done.

bfredl commented:
Thanks for doing this. Good choices for error handling and `print()` support :+1:

erw7 committed:
feat(lua): add vim.inspect,shared in thread

erw7 committed:
feat(lua): add vim.NIL,_empty_dict_mt in thread

erw7 committed:
refactor(lua): change to use helper func for initialization

bfredl commented on src/nvim/lua/executor.c:
```diff
@@ -1043,12 +1234,19 @@ LuaRef nlua_ref(lua_State *lstate, int index)
   LuaRef ref = luaL_ref(lstate, LUA_REGISTRYINDEX);
   if (ref > 0) {
-    nlua_refcount++;
+    nlua_inc_refcount(lstate);
```
This much extra complexity for a what is supposed to be a primitive operation is a bit unfortunate. But AFAIK, references tracked by nvim is only used by editor state keeping refs to lua, so these won't be used by lua threads anyway. So for now we could just revert this and document `nlua_ref` is not to be used for threads. 

 But if we do later on, we should add a single userdata for thread-local state which can passed as a _c pointer_ to internal helper functions so there only will be one register (read) access per enter c code from lua to get this struct.

erw7 committed:
refactor(lua): remove reference related code from thread

bfredl commented on src/nvim/lua/executor.c:
```diff
@@ -520,51 +621,18 @@ static int nlua_state_init(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
   lua_setfield(lstate, LUA_REGISTRYINDEX, "mpack.empty_dict");
   lua_setfield(lstate, -2, "_empty_dict_mt");
 
```
we could move `vim.NIL` and `vim._empty_dict_mt` also into `nlua_common_vim_init`. just place the `nlua_empty_dict_ref` and `nlua_nil_ref` assigments within `if (!is_thread) { ... }` conditions.

erw7 replied:
I've moved them and `vim.loop`.

erw7 committed:
refactor(lua): move common parts into nlua_common_vim_init

erw7 commented:
A patch (luvit/luv/pull/563) to run the thread entry function itself in protected mode has been merged into luv. It may be better to merge this after the modified luv is released.

bfredl commented:
Let's wait for a comment on https://github.com/luvit/luv/issues/564 as well :)

erw7 committed:
build(third-party): upgrade libluv: 1.40.0-0 => 1.42.0-0

Changes: https://github.com/luvit/luv/compare/1.40.0-0...1.42.0-0

erw7 committed:
fix(lua): change all luv functions to run in protected mode

brcolow commented on src/nvim/lua/executor.c:
```diff
@@ -397,6 +480,105 @@ static int nlua_wait(lua_State *lstate)
+static void nlua_common_vim_init(lua_State *lstate, bool is_thread)
+  FUNC_ATTR_NONNULL_ARG(1)
+{
+  // vim.is_therad
```
Typo: vim.istherad => vim.isthread

erw7 committed:
feat(lua): add vim.json in thread

bfredl commented on src/nvim/lua/executor.c:
```diff
@@ -397,6 +488,154 @@ static int nlua_wait(lua_State *lstate)
+static void nlua_common_vim_init(lua_State *lstate, bool is_thread)
+  FUNC_ATTR_NONNULL_ARG(1)
+{
+  nlua_ref_state_t *ref_state = xcalloc(1, sizeof(*ref_state));
+  nlua_init_ref_state(ref_state);
+  lua_pushlightuserdata(lstate, ref_state);
```
a lightuserdata that "owns" an allocation is an anti-pattern. use a full userdata to store the `nlua_ref_state_t` directly instead.

erw7 committed:
test(lua): add test for thread

erw7 committed:
feat(lua): add vim.api.nvim__get_runtime to thread

erw7 committed:
build(deps): bump luv to 53e471da71a01e94f557d4ca0abda5a535205bb7

erw7 committed:
feat(lua): change to allow loading from rtp on thread

erw7 committed:
refactor(lua): change to use mutex for runtime path

bfredl commented:
I'd say we don't want to introduce multi-thread features just at a feature freeze. It is better to do this just after 0.6 release then we have the entire 0.7 cycle for early testers and do design re-reiterations as needed.

mjlbach commented:
I'm not sure if the default options for cjson are thread safe, specifically the encode_keep_buffer option which is on by default https://github.com/mpx/lua-cjson/blob/e8972ac754788d3ef10a57a36016d6c3e85ba20d/lua_cjson.c#L309-L328

erw7 committed:
fix(lua): change fpconv_init to be called only once in main thread

erw7 commented:
> I'm not sure if the default options for cjson are thread safe, specifically the encode_keep_buffer option which is on by default https://github.com/mpx/lua-cjson/blob/e8972ac754788d3ef10a57a36016d6c3e85ba20d/lua_cjson.c#L309-L328

The `json_config_t *cfg` returned by `json_arg_init()` in the relevant section is the user data obtained from the upvalue of `lstate`, which is created for each thread. Also, `cfg->encode_buf.buf` when `cfg->encode_keep_buffer` is true is not a problem since it is allocated by `malloc` in `strbuf_init()`.

What is not thread safe are global variables and static variables in functions. Even access to these should be safe as long as all concurrent accesses are read operations (I'm not sure if it's really safe for all architectures, but I think it's okay).

erw7 committed:
test(lua): add more test for thread

erw7 committed:
fix(main): change order of fs_init calls

Fix the problem that fs_init() was not called before os_open() was called
in init_startuptime().

bfredl commented:
I am gonna work on rebase this and some minor changes (like reduce the overhead of using refs on the main thread). Then we can get this merged pretty soon, and work on improving thread support on the master branch during the 0.7 cycle.

bfredl committed:
fixup: mongled the rebase

bfredl committed:
refactor(lua): use references directly on main thread

bfredl committed:
fubbit

clason committed:
Merge branch 'master' into add-vim-tbl-to-thread

clason commented:
Ugh, I didn't realize Github would merge master into that PR (instead of rebasing, like a sane person). Please drop that commit into a bin and burn it with fire. @bfredl

vhyrro commented:
Hey uh any info on what happened here? :p

Seems like erw7 closed every single neovim PR he's ever made :(

vhyrro commented:
Hey uh any info on what happened here? :p

Seems like erw7 closed every single neovim PR he's ever made :(
